### PR TITLE
fix(PluginResolver.ts): add ./node_modules to module load path

### PR DESCRIPTION
**Pull request by dogatana**
change the order to resolve a package location as follows

- try ./node_modules first (added)
- try baseDir next (original)

This is a solution for #133

### CODE REVIEW

File: packages/honkit/src/plugins/PluginResolver.ts [hunk 1]
```diff
@@ -47,12 +47,22 @@ class PluginResolver {
...
         // In sometimes, HonKit package has not main field - so search package.json
-        const pkgPath =
-            tryResolve(path.join(baseDir, honkitFullPackageName, "/package.json")) ||
-            tryResolve(path.join(baseDir, gitbookFullPackageName, "/package.json")) ||
-            tryResolve(path.join(baseDir, packageName, "/package.json")) ||
-            (SPECIAL_PACKAGE_NAME.includes(packageName) &&
-                tryResolve(path.join(baseDir, honkitScopePackageName, "/package.json")));
+
+        // search ./node_modules first, then baseDir
+        for (const dir of [path.join(process.cwd(), "node_modules"), baseDir]) {
```
Comment (azu) [inline packages/honkit/src/plugins/PluginResolver.ts:52]: I do not realize what this logic does.
We need to write tests as expected behavior like https://github.com/textlint/textlint/blob/master/packages/textlint/test/textlint-module-resolver/textlint-module-resolver-test.js.  

Currently, `baseDir` always to be `""`.
https://github.com/honkit/honkit/blob/ed8ec44549c4ed9253c56dc9f24fda1f34449953/packages/honkit/src/plugins/loadForBook.ts#L22

So, It is just `require.resolve(pkgName)`.

- https://nodejs.org/api/modules.html#modules_all_together

**Review by azu — COMMENTED** [review]
Probably, we need to define the loading logic for the plugin.

The plugin should be loaded from ... ?

- current directory(This PR)
- Honkit installed directory(Current maybe)
- config directory like ESLint
- configuration like `--cwd` flag
- other

Related works: https://github.com/eslint/eslint/blob/b5c8d7389bb2c5d4eae850b866832d099187818b/docs/user-guide/migrating-to-7.0.0.md#plugin-loading-change

For example, if loading from the current directory, the global installed `honkit` command looks into the local directory at first.
Is it ok?

**Comment by dogatana** [comment]
To make following order clear, modify PluginResolver class and lordForBook.ts 

> - currentt directory(This PR)
> - Honkit installed directory(Current maybe)

I hope this can be applied for other cases by change the config.

> - config directory like ESLint
> - configuration like --cwd flag

On the other hand,

> We need to write tests as expected behavior

I can understand it, but it's hard for me.
I have no idea how to setup try-resolve stub, and it seems that further modification will be necessary for test.

File: packages/honkit/src/plugins/PluginResolver.ts (hunk 1 above)
Comment (dogatana) [reply packages/honkit/src/plugins/PluginResolver.ts]: It's a quick way but too dirty..
I changed both PluginResolver.ts and loadForBook.ts to make it clear.

**Review by azu — CHANGES_REQUESTED** [review]
>  I have no idea how to setup try-resolve stub, and it seems that further modification will be necessary for test.

Good point. It is dificulet to test module's directory order.

We can add test case for `baseDirectories` and `resolvePluginPackageName` method behavior.
`PluginResolver` receive `baseDirectories` as options.
It means that the test case can pass the test's fixture directory to `PluginResolver`.

For example, 

- fixtures/
  - modules/
     - example
  - another_modules/
     - example
     - another_example

```ts
const resolver = PluginResolver({
  baseDirectories: [path.join(__dirname, "fixtures/modules")]
});
const fooPath = resolver.resolvePluginPackageName("example"); // it will resolve path.join(__dirname, "fixtures/modules/example/package.json");
assert.strictEqual(fooPath, path.join(__dirname, "fixtures/modules/example"));
```

and 

```ts
const resolver = PluginResolver({
  baseDirectories: [path.join(__dirname, "fixtures/modules"), path.join(__dirname, "fixtures/another_modules")]
});
const fooPath = resolver.resolvePluginPackageName("example");
assert.strictEqual(fooPath, path.join(__dirname, "fixtures/modules/example"));
const anotherFooPath = resolver.resolvePluginPackageName("another_example");
assert.strictEqual(anotherFooPath, path.join(__dirname, "fixtures/another_modules/another_example"));

```

and etc...

similar tests:

https://github.com/textlint/textlint/blob/master/packages/textlint/test/textlint-module-resolver/textlint-module-resolver-test.js
https://github.com/textlint/textlint/tree/master/packages/textlint/test/textlint-module-resolver

File: packages/honkit/src/plugins/loadForBook.ts [hunk 2]
```diff
@@ -19,11 +20,13 @@ function loadForBook(book) {
      */
     const requirements = listDepsForBook(book);
 
-    const pluginResolver = new PluginResolver();
+    const pluginResolver = new PluginResolver({
+        baseDirectories: [path.join(process.cwd(), "node_modules")],
```
[the remark this answers is not in the document]
Comment (azu) [reply packages/honkit/src/plugins/loadForBook.ts:24]: I don't really have much confidence about `require.resolve.paths` usage.
https://nodejs.org/api/modules.html#modules_require_resolve_request_options

File: packages/honkit/src/plugins/PluginResolver.ts [hunk 3]
```diff
@@ -24,40 +24,54 @@ const SPECIAL_PACKAGE_NAME = [
...
     }
 
     /**
      * Take package name, and return path to module.
      * @param {string} packageName
      * @returns {string} return path to module
      */
-    resolvePluginPackageName(packageName) {
-        const baseDir = this.baseDirectory;
+    resolvePluginPackageName(packageName, logger) {
```
[the remark this answers is not in the document]
Comment (dogatana) [reply packages/honkit/src/plugins/PluginResolver.ts]: I've changed.

**Comment by dogatana** [comment]
I drop this.
The reason is the same as #132.

At last, I've created a theme plugin that solves #132 and #135.
It replaces "theme-default"  and works well with standard, npm-installed, honkit so far.
Thank you very much for your advice.

image

**Pull request closed without merging**